File: eco-kernel-cpp/src/eco/Process.cpp
//===- Process.cpp - Process kernel module implementation -----------------===//

#include "Process.hpp"
#include "KernelHelpers.hpp"
#include <cstdlib>
#include <string>
#include <unordered_map>
#include <vector>
#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>

namespace Eco::Kernel::Process {

// Map from child PID to pipe fd (stdin write end) if applicable.
static std::unordered_map<int64_t, int> s_streamHandles;

uint64_t exit(int64_t code) {
    ::exit(static_cast<int>(code));
    // Never returns.
    return 0;
}

uint64_t spawn(uint64_t cmd, uint64_t args) {
    std::string cmdStr = toString(cmd);
    std::vector<std::string> argStrs = listToStringVector(args);

    // Build argv array for execvp.
    std::vector<char*> argv;
    argv.push_back(const_cast<char*>(cmdStr.c_str()));
    for (auto& a : argStrs) {
        argv.push_back(const_cast<char*>(a.c_str()));
    }
    argv.push_back(nullptr);

    pid_t pid = fork();
    if (pid < 0) {
        return taskFailString("fork failed");
    }
    if (pid == 0) {
        // Child process.
        execvp(argv[0], argv.data());
        ::_exit(127);
    }
    return taskSucceedInt(static_cast<int64_t>(pid));
}

uint64_t spawnProcess(uint64_t cmd, uint64_t args,
                      uint64_t stdin_, uint64_t stdout_, uint64_t stderr_) {
    std::string cmdStr = toString(cmd);
    std::vector<std::string> argStrs = listToStringVector(args);
    std::string stdinCfg = toString(stdin_);
    std::string stdoutCfg = toString(stdout_);
    std::string stderrCfg = toString(stderr_);

    int stdinPipe[2] = {-1, -1};
    bool pipeStdin = (stdinCfg == "pipe");
    if (pipeStdin) {
        if (pipe(stdinPipe) < 0) {
            return taskFailString("pipe failed");
        }
    }

    // Build argv.
    std::vector<char*> argv;
    argv.push_back(const_cast<char*>(cmdStr.c_str()));
    for (auto& a : argStrs) {
        argv.push_back(const_cast<char*>(a.c_str()));
    }
    argv.push_back(nullptr);

    pid_t pid = fork();
    if (pid < 0) {
        if (pipeStdin) {
            ::close(stdinPipe[0]);
            ::close(stdinPipe[1]);
        }
        return taskFailString("fork failed");
    }

    if (pid == 0) {
        // Child process.
        if (pipeStdin) {
            dup2(stdinPipe[0], STDIN_FILENO);
            ::close(stdinPipe[0]);
            ::close(stdinPipe[1]);
        }
        // stdout/stderr: "inherit" is the default (do nothing).
        execvp(argv[0], argv.data());
        ::_exit(127);
    }

    // Parent process.
    if (pipeStdin) {
        ::close(stdinPipe[0]); // Close read end in parent.
    }

    // Build result record: { stdinHandle : Maybe Int, processHandle : Int }
    // Layout (unboxed-first):
    //   Slot 0: processHandle (Int, unboxed)
    //   Slot 1: stdinHandle (Maybe Int, boxed)
    //   Bitmap: 0b01
    using namespace Elm::alloc;
    std::vector<Unboxable> fields(2);
    fields[0].i = static_cast<int64_t>(pid);

    if (pipeStdin) {
        int64_t handleId = stdinPipe[1];
        s_streamHandles[handleId] = stdinPipe[1];
        HPointer handleInt = allocInt(handleId);
        fields[1].p = just(boxed(handleInt), true);
    } else {
        fields[1].p = nothing();
    }

    HPointer rec = record(fields, 0b01);
    return taskSucceed(rec);
}

uint64_t wait(uint64_t handle) {
    int64_t pid = static_cast<int64_t>(handle);
    int status = 0;
    waitpid(static_cast<pid_t>(pid), &status, 0);
    int exitCode = WIFEXITED(status) ? WEXITSTATUS(status) : 1;
    return taskSucceedInt(exitCode);
}

} // namespace Eco::Kernel::Process

File: eco-kernel-cpp/src/Eco/Kernel/Process.js
/*
import Eco.Kernel.Scheduler exposing (succeed, fail, binding)
import Eco.Kernel.Utils exposing (Tuple2)
import Elm.Kernel.List exposing (toArray)
import Maybe exposing (Just, Nothing)
*/

var _Process_children = {};
var _Process_streamHandles = {};
var _Process_nextStreamHandle = 10000;

var _Process_exit = function(code) {
    return __Scheduler_binding(function(callback) {
        process.exit(code);
    });
};

var _Process_spawn = F2(function(cmd, args) {
    return __Scheduler_binding(function(callback) {
        try {
            var child_process = require('child_process');
            var child = child_process.spawn(cmd, __List_toArray(args),
                { stdio: ['inherit', 'inherit', 'inherit'] });
            _Process_children[child.pid] = child;
            callback(__Scheduler_succeed(child.pid));
        } catch (e) {
            callback(__Scheduler_fail(e.message));
        }
    });
});

var _Process_spawnProcess = F5(function(cmd, args, stdin, stdout, stderr) {
    return __Scheduler_binding(function(callback) {
        try {
            var child_process = require('child_process');
            var child = child_process.spawn(cmd, __List_toArray(args),
                { stdio: [stdin, stdout, stderr] });
            _Process_children[child.pid] = child;
            var stdinHandle;
            if (child.stdin) {
                var handleId = _Process_nextStreamHandle++;
                _Process_streamHandles[handleId] = child.stdin;
                stdinHandle = __Maybe_Just(handleId);
            } else {
                stdinHandle = __Maybe_Nothing;
            }
            callback(__Scheduler_succeed(
                __Utils_Tuple2(stdinHandle, child.pid)
            ));
        } catch (e) {
            callback(__Scheduler_fail(e.message));
        }
    });
});

var _Process_wait = function(handle) {
    return __Scheduler_binding(function(callback) {
        var child = _Process_children[handle];
        if (!child) {
            callback(__Scheduler_succeed(0));
            return;
        }
        if (child.exitCode !== null) {
            delete _Process_children[handle];
            callback(__Scheduler_succeed(child.exitCode));
            return;
        }
        child.on('exit', function(code) {
            delete _Process_children[handle];
            callback(__Scheduler_succeed(code || 0));
        });
    });
};
